processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java

```diff
@@ -312,12 +314,99 @@ public RowIterator(GroupByQuery query, Cursor cursor, ByteBuffer metricsBuffer,
       dimensionSpecs = query.getDimensions();
       dimensions = Lists.newArrayListWithExpectedSize(dimensionSpecs.size());
       dimNames = Lists.newArrayListWithExpectedSize(dimensionSpecs.size());
+
       for (int i = 0; i < dimensionSpecs.size(); ++i) {
         final DimensionSpec dimSpec = dimensionSpecs.get(i);
-        final DimensionSelector selector = cursor.makeDimensionSelector(dimSpec.getDimension());
-        if (selector != null) {
-          dimensions.add(selector);
+        final String dimension = dimSpec.getDimension();
+        if(dimension.equals(Column.TIME_COLUMN_NAME)) {
+          if(dimSpec.getExtractionFn() == null) {
+            throw new UnsupportedOperationException("Cannot group by time without time extraction function");
+          }
+          // use special DimSelector for projected time columns
+          // assumes timestamps are grouped
+          dimensions.add(
+              new DimensionSelector()
+              {
+                private final LongColumnSelector timeSelector = cursor.makeLongColumnSelector(dimension);
+                private final Map<Integer, String> timeValues = Maps.newHashMap();
+                private final ExtractionFn timeExtractionFn = dimSpec.getExtractionFn();
+                private String currentValue = null;
+                private long currentTimestamp = Long.MIN_VALUE;
+                private int index = -1;
+
+                @Override
+                public IndexedInts getRow()
+                {
+                  return new IndexedInts()
+                  {
+                    @Override
+                    public int size()
+                    {
+                      return 1;
+                    }
+
+                    @Override
+                    public int get(int index)
+                    {
+                      return getTimeBucketIndex();
+                    }
+
+                    protected Integer getTimeBucketIndex()
+                    {
+                      final long timestamp = timeSelector.get();
```

Comment:
I think you can move this logic to exist before `return new IndexedInts()...` and effectively just generate an `final int index` which is passed into the anonymous class.

Hopefully that will clarify what is happening in the code as well.